docs/architecture/adr-020-protobuf-transaction-encoding.md

```diff
@@ -137,38 +170,97 @@ that account fields can be retrieved for signing.
-module level and a no-arg constructor function should be passed around to
-required CLI and REST client commands.
+If a module needs to work with `sdk.Msg`s that use interface types, those interface
+types should be dealt with using `google.protobuf.Any` for signing and a `oneof`
+at the app-level for encoding.
+
+Using `google.protobuf.Any` for signing will allow client libraries to create
+reusable code for dealing with interface types that doesn't require regenerating
+protobuf client libraries for every chain. Using `oneof`s at the encoding level
+saves space in the Tendermint block store.
+
+Modules should define a generic `sdk.Msg` for interface types at the module
+level that uses `Any`. Ex:
+
+```go
+// x/gov/types/types.proto
+message MsgSubmitAnyProposal {
+  MsgSubmitProposalBase base = 1;
+  google.protobuf.Any content = 2;
+}
+```
 
-For example, in `x/gov`, `Content` is an interface type, so `MsgSubmitProposalI`
-should also be an interface and implement setter methods:
+Apps should define an app-specific `sdk.Msg` that encodes the concrete types
+it supports using an `oneof`:
 
 ```go
-// x/gov/types/msgs.go
-type MsgSubmitProposalI interface {
-	sdk.Msg
+// myapp/types/types.proto
+message MsgSubmitProposal {
+  MsgSubmitProposalBase base = 1;
+  MyAppContent content = 1;
+}
 
-	GetContent() Content
-    // SetContent returns an error if the underlying oneof does not support
-    // the concrete Content passed in
-	SetContent(Content) error
+message MyAppContent {
+  oneof sum {
+    TextProposal text = 1;
+    SomeOtherProposal = 2;
+  }
+}
+```
 
-	GetInitialDeposit() sdk.Coins
-	SetInitialDeposit(sdk.Coins)
+**Client libraries should always sign transactions using the generic module-level
+`sdk.Msg` that uses `Any`.** Convenience gRPC/REST methods will be provided
+that allow the generic module-level `Msg` to be used for transaction
+submission, converting it to the app-level encoding `Msg` using the `oneof`.
 
-	GetProposer() sdk.AccAddress
-	SetProposer(sdk.AccAddress)
+In order to smoothly allow for conversion between _signing_ and _encoding_
+`Msg`s, modules that use interface types should implement the following
+interface:
+
+```go
+type InterfaceMsgEncoder {
+  GetSigningMsg(sdk.Msg) (sdk.Msg, error)
+  GetEncodingMsg(sdk.Msg) (sdk.Msg, error)
 }
 ```
 
-Note that the implementation of `MsgSubmitProposalI` can be simplified by
-using an embedded base struct which implements most of that interface - in this
-case `MsgSubmitProposalBase`.
+`GetSigningMsg` should type switch over the `sdk.Msg`s that use interfaces and
+convert `Msg`s in the encoding format (that use `oneof`)
+to the signing version (that uses `Any`). `GetEncodingMsg` should
+convert `Msg`s used for signing (using `Any`) to those used for encoding
+(using `oneof`). Ex:
+
+```go
+// x/gov//module.go
+var _ InterfaceMsgEncoder = AppModule{}
+
+type AppModule struct {
+  ...
+  newEncodingMsg fn(MsgSubmitProposalBase, Content) (MsgSubmitProposalI, error)
```

Comment:
Why does this function have to be part of the `AppModule`? How/when is it provided? Is its implementation dynamic?

Reply:
No, it doesn't. It just seemed like a convenient place because we can add `InterfaceMsgEncoder` as an optional interface for `AppModules`, then iterate over all of the modules and figure out which ones implement it via reflection and do the encoding. I haven't worked out all the details but it's needed whenever transactions are generated, likely mostly in `client/tx`

Reply:
I have been working on the implementation for this in another branch and can post a draft PR. Once I work through all of that, I'll likely have a better sense.

Reply:
Ok, yeah I'm just getting at the fact that if it is not app-specific or doesn't change between clients/apps, we should just have it hard-coded implementation if possible. If not, that's fine.

Reply:
I'm not quite sure what you mean. Could you explain a bit more?

Reply:
I mean not have it be an arg at all. Just code the logic in `GetSigningMsg`. However, you seem to have a good grasp on this, so just go with you believe is best.

Reply:
Well `GetSigningMsg` would be a module-level function that takes an app-level parameter. So somewhere the app needs to pass it into the module. I actually think this will be simpler than what we previously had in #5942 where the `newMsgFn` argument gets passed around to CLI methods. Those CLI methods can just submit the signing msg and `GetEncodingMsg` takes care of the rest. Anyway, I think it will be okay. We can review in the implementation PR.